src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,227 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "sol-flow/netctl.h"
+#include "sol-flow-internal.h"
+
+#include <sol-netctl.h>
+#include <sol-str-table.h>
+#include <sol-util-internal.h>
+#include <errno.h>
+
+struct list_service_data {
+    struct sol_ptr_vector service_list;
+    struct sol_flow_node *node;
+};
+
+struct connection_service_data {
+    struct sol_flow_node *node;
+    char *ap_name;
+    bool connect;
+};
+
+void
+network_service_cb(void *data, const struct sol_netctl_service *service)
+{
+    struct list_service_data *mdata = data;
+    struct sol_flow_node *node = mdata->node;
+    int r;
+    const char *name;
+
+    if (sol_netctl_service_get_state(service) ==
+        SOL_NETCTL_SERVICE_STATE_REMOVE) {
+        sol_ptr_vector_del_element(&mdata->service_list, service);
+        return;
+    }
+
+    name = sol_netctl_service_get_name(service);
+    r = sol_flow_send_string_packet(node,
+        SOL_FLOW_NODE_TYPE_NETCTL_LIST_SERVICES__OUT__OUT,
+        name);
+    if (r < 0)
+        SOL_WRN("Failed to send network service name: %s", name);
+
+    sol_ptr_vector_append(&mdata->service_list, service);
+}
+
+void
+network_connection_cb(void *data, const struct sol_netctl_service *service)
+{
+    struct connection_service_data *mdata = data;
+    struct sol_flow_node *node = mdata->node;
+    enum sol_netctl_service_state state;
+    int r;
+    const char *name;
+
+    state = sol_netctl_service_get_state(service);
+    name = sol_netctl_service_get_name(service);
+
+    if (name && strcmp(name, mdata->ap_name) == 0)
+        if (state == SOL_NETCTL_SERVICE_STATE_IDLE && mdata->connect) {
+            sol_netctl_service_connect(service);
+            sol_flow_send_string_packet(node,
+                SOL_FLOW_NODE_TYPE_NETCTL_CONNECTION__OUT__OUT,
+                "Connect AP!");
+        } else if ((state == SOL_NETCTL_SERVICE_STATE_READY) && (!mdata->connect)) {
+            sol_netctl_service_disconnect(service);
+            sol_flow_send_string_packet(node,
+                SOL_FLOW_NODE_TYPE_NETCTL_CONNECTION__OUT__OUT,
+                "Disconnect AP!");
+        }
+}
+
+void
+error_cb(void *data, const struct sol_netctl_service *service,
+    unsigned int error)
+{
+    const char *name;
+
+    name = sol_netctl_service_get_name(service);
+
+    if (name)
+        SOL_ERR("Service %s error is %d", name, error);
+}
+
+static int
+open_list_services(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    static const char *errmsg = "Could not add service monitor";
+    struct list_service_data *mdata = data;
+
+    mdata->node = node;
+
+    sol_ptr_vector_init(&mdata->service_list);
+
+    if (sol_netctl_add_service_monitor(network_service_cb, mdata))
+        goto error;
+
```

Comment:
here you'd get `sol_netctl_get_services()` and emit them all in the `ADDED` port. (see my other comment)